[build] rewrite #21707 so that it doesn't break cross-compilation

finagolfin opened the pull request:
Passing swift_core_private_link_libraries to PRIVATE_LINK_LIBRARIES only works either when building a single host SDK alone or when needed by all SDKs, so move all that configuration to a swift_core_private_libraries() function that's called from add_swift_target_library() for each SDK/arch instead. Similarly, call a swift_runtime_static_libraries() function to split off libraries for the static stdlib with linux. Finally, remove the last uses of `SWIFT_CONFIGURED_SDKS` for anything more than checking, using `SWIFT_SDKS` instead, and move the static-stdlib linker file generation to the `stdlib/` directory.

I ran into this when trying out the Android cross-compilation commands in #29439 with Swift 5.1.4 built from source on linux x86_64, which I fixed back then with a much simpler version of this pull. It broke because I wasn't building `SWIFT_PATH_TO_LIBICU_BUILD`, instead linking the linux SDK against the system libicu package and the Android SDK against a pre-built libicu for Android.

That would make the prior CMake config from #21707 try to link the Android swiftCore against the linux libicu, ie the libicu of the primary variant SDK. The Android CI didn't detect this because [it builds libicu from source](https://github.com/apple/swift/blob/0a46bd457be2fc16c8efcd64268b06580ba87da3/utils/build-presets.ini#L882), not using the system libicu for linux like I did. The rest of these non-Android SDKs aren't cross-compiled on any CI, so the cross-compilation regression wasn't found.

`SWIFT_PRIMARY_VARIANT_SDK` seems to be a meaningless variable, which AFAICT means "usually the host SDK but not always." It should either be defined much more precisely, or removed altogether and replaced with checking if an sdk is in `SWIFT_SDKS` or something.

This pull was tested by back-porting to that Swift 5.1.4 and now 5.2.1 cross-compilation setup and running the same commands to produce a working executable on Android.

buttaface commented on stdlib/public/core/CMakeLists.txt:
```diff
@@ -246,34 +246,42 @@ else()
+
+  if(sdk STREQUAL ANDROID)
+    # workaround for libatomic needing to be after object files for swiftCore.so
+    list(APPEND private_link_libraries atomic)
```
This library isn't being linked here by the Android CI currently, maybe it isn't needed?

buttaface replied:
@CodaFi, you reviewed the pull when this was put in, #9869, appreciate a review on this whole pull too when you have the time.

@drexin, can these two libatomic dependencies be removed too, like in #30155?

drexin replied:
I‘m not sure about Android. Can you remove them and see if it still compiles?

buttaface replied:
Yep, I removed the libatomic dependency here and in `_add_variant_link_flags` and everything still compiled on an Android host. I guess this dependency can be removed for Android and BSD too?

buttaface commented on stdlib/public/core/CMakeLists.txt:
```diff
@@ -246,34 +246,42 @@ else()
+    # workaround for libatomic needing to be after object files for swiftCore.so
+    list(APPEND private_link_libraries atomic)
+  elseif(sdk STREQUAL FREEBSD)
+    find_library(EXECINFO_LIBRARY execinfo)
```
I don't know if `find_library` will work when cross-compiling, so this will need to be pulled out like `CYGWIN` below if not.

compnerd replied:
`find_library` is completely capable of working with cross-compilation, please read up on CMake's cross-compilation support, its pretty well documented.

buttaface replied:
I know, I subsequently used it [when cross-compiling the Swift stdlib for Android](https://github.com/termux/termux-packages/pull/5169/files#diff-bd95e8ae22237283c1199931c04d509aR12). Anyway, nobody is likely to ever cross-compile the Swift stdlib to BSD, so this point is moot.

buttaface commented on stdlib/public/core/CMakeLists.txt:
```diff
@@ -246,34 +246,42 @@ else()
+
+if(SWIFT_HOST_VARIANT_SDK STREQUAL CYGWIN)
   # TODO(compnerd) cache this variable to permit re-configuration
   execute_process(COMMAND "cygpath" "-u" "$ENV{SYSTEMROOT}" OUTPUT_VARIABLE ENV_SYSTEMROOT)
```
I think this will only work when natively compiling, so I pulled it back out.

buttaface commented:
Alright, @gottesmm, review requested.

finagolfin committed:
[build] rewrite #21707 and swiftCore/swiftRuntime CMake config, so that it doesn't break
cross-compilation

Passing swift_core_private_link_libraries to PRIVATE_LINK_LIBRARIES only works either when
building a single host SDK alone or when needed by all SDKs, so move all that configuration
to a swift_core_private_libraries() function that's called from add_swift_target_library()
for each SDK/arch instead. Also, generate a static-executable-args.lnk for each ELF SDK,
rather than just for linux if it is the host SDK.

buttaface commented on cmake/modules/AddSwift.cmake:
```diff
@@ -417,10 +417,10 @@ function(_add_variant_link_flags)
     list(APPEND result "-Wl,-Bsymbolic")
   elseif("${LFLAGS_SDK}" STREQUAL "ANDROID")
-    list(APPEND link_libraries "dl" "log" "atomic")
+    list(APPEND link_libraries "dl" "log")
```
@drexin, updated this pull to remove remaining libatomic dependencies.

buttaface commented on cmake/modules/AddSwift.cmake:
```diff
@@ -417,22 +417,17 @@ function(_add_variant_link_flags)
-    # link against the ICU libraries
-    list(APPEND link_libraries
-      ${SWIFT_ANDROID_${LFLAGS_ARCH}_ICU_I18N}
-      ${SWIFT_ANDROID_${LFLAGS_ARCH}_ICU_UC})
```
I noticed this was adding the full libicu path twice for Android now that it's set directly in swiftCore, so removed this.

buttaface commented:
After thinking about it some more, I realized that the entire previous approach of setting `swift_core_private_link_libraries` is flawed, because [`add_swift_target_library()` sets anything passed to `PRIVATE_LINK_LIBRARIES` for _every_ sdk and arch](https://github.com/apple/swift/blob/fc67fc74fb97d4708686b23f36a0370fd8c2f37a/stdlib/cmake/modules/AddSwiftStdlib.cmake#L1429)! This only worked so far because almost nobody is cross-compiling the stdlib, particularly for non-Apple platforms, but as soon as I tried cross-compiling the Android stdlib with a slightly different config, it broke. So I removed `swift_core_private_link_libraries` altogether and now set everything specifically for each sdk/arch through the new function `swift_core_private_libraries()`.

By removing the use of `PRIVATE_LINK_LIBRARIES` here, I looked and it is no longer used in this open source repo. I think that argument to `add_swift_target_library()` should be removed altogether, as it's only useful when passing in a library that will be used by _every_ SDK/arch combo, which there isn't a need for atm. But I don't know how else people are using that function argument, perhaps in closed-source code, so I need a go-ahead on that.

@drodriguez, let me know what you think of this pull.

drodriguez commented:
I think this is going to break, and more than Android. I just tested Android ARMv7 and I have 46 new failures (I am verifying, and I will test Android AArch64 as soon as I can). Many of the failures are a missing symbol swift_addNewDSOImage.

Let's not merge this until I figure this out. Starting the tests to see if the Linux tests shows the same problem.

@swift-ci please test

drodriguez commented:
So I get the following 46 also in AArch64. I guess these are going to fail in the CI machines too.

```
Failing Tests (46):
    Swift(android-aarch64) :: Interpreter/use_public_var_private_setter.swift
    Swift(android-aarch64) :: multifile/error-type/two-modules/main.swift
    Swift(android-aarch64) :: multifile/error-type/one-module/main.swift
    Swift(android-aarch64) :: multifile/synthesized-accessors/one-module-public/main.swift
    Swift(android-aarch64) :: multifile/constant-struct-with-padding/main.swift
    Swift(android-aarch64) :: multifile/synthesized-accessors/one-module-internal/main.swift
    Swift(android-aarch64) :: multifile/constant-tuple-with-padding/main.swift
    Swift(android-aarch64) :: Driver/response-file.swift
    Swift(android-aarch64) :: Driver/opt-remark.swift
    Swift(android-aarch64) :: Driver/loaded_module_trace.swift
    Swift(android-aarch64) :: Driver/loaded_module_trace_swiftinterface.swift
    Swift(android-aarch64) :: IRGen/vtable_symbol_linkage.swift
    Swift(android-aarch64) :: Misc/stats_dir_plausible_maxrss.swift
    Swift(android-aarch64) :: ModuleInterface/option-preservation.swift
    Swift(android-aarch64) :: SILOptimizer/alwaysemitintoclient.swift
    Swift(android-aarch64) :: TypeDecoder/nominal_types.swift
    Swift(android-aarch64) :: TypeDecoder/existentials.swift
    Swift(android-aarch64) :: TypeDecoder/generic_local_types.swift
    Swift(android-aarch64) :: TypeDecoder/structural_types.swift
    Swift(android-aarch64) :: TypeDecoder/dynamic_self.swift
    Swift(android-aarch64) :: TypeDecoder/extensions.swift
    Swift(android-aarch64) :: TypeDecoder/builtins.swift
    Swift(android-aarch64) :: TypeDecoder/reference_storage.swift
    Swift(android-aarch64) :: TypeDecoder/lowered_function_types.swift
    Swift(android-aarch64) :: TypeDecoder/generic_typealias.swift
    Swift(android-aarch64) :: TypeDecoder/opaque_return_type.swift
    Swift(android-aarch64) :: TypeDecoder/generics.swift
    Swift(android-aarch64) :: TypeDecoder/lowered_metatypes.swift
    Swift(android-aarch64) :: TypeDecoder/local_types.swift
    Swift(android-aarch64) :: TypeDecoder/sugar.swift
    Swift(android-aarch64) :: TypeDecoder/typealias.swift
    Swift(android-aarch64) :: multifile/default-arguments/two-modules/main.swift
    Swift(android-aarch64) :: multifile/extensions/two-modules/main.swift
    Swift(android-aarch64) :: multifile/class-layout/final-stored-property/main.swift
    Swift(android-aarch64) :: multifile/synthesized-accessors/two-modules/main.swift
    Swift(android-aarch64) :: multifile/default-arguments/one-module/main.swift
    Swift(android-aarch64) :: multifile/typealias/two-modules/main.swift
    Swift(android-aarch64) :: multifile/typealias/one-module/main.swift
    Swift(android-aarch64) :: PlaygroundTransform/iuo.swift
    Swift(android-aarch64) :: LinkerSections/function_sections.swift
    Swift(android-aarch64) :: Driver/multi-threaded.swift
    Swift(android-aarch64) :: Misc/stats_dir_profiler.swift
    Swift(android-aarch64) :: Reflection/typeref_lowering.swift
    Swift(android-aarch64) :: Driver/opt-record.swift
    Swift(android-aarch64) :: Misc/stats_dir_tracer.swift
    Swift(android-aarch64) :: Reflection/typeref_decoding.swift
```

Example of a failing test:

```
FAIL: Swift(android-aarch64) :: PlaygroundTransform/iuo.swift (1 of 1)
******************** TEST 'Swift(android-aarch64) :: PlaygroundTransform/iuo.swift' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf "/home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp" && mkdir -p "/home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp"
: 'RUN: at line 2';   cp /home/danielrodriguez/swift-source/swift/test/PlaygroundTransform/iuo.swift /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp/main.swift
: 'RUN: at line 3';   /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/bin/swiftc -target aarch64-unknown-linux-android -Xcc --sysroot=/home/danielrodriguez/android-ndk-r18/platforms/android-21/arch-arm64 -Xclang-linker --sysroot=/home/danielrodriguez/android-ndk-r18/platforms/android-21/arch-arm64 -tools-directory /home/danielrodriguez/android-ndk-r18/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/aarch64-linux-android/bin -I /home/danielrodriguez/android-ndk-r18/sysroot/usr/include -I /home/danielrodriguez/android-ndk-r18/sysroot/usr/include/aarch64-linux-android -L /home/danielrodriguez/android-ndk-r18/sources/cxx-stl/llvm-libc++/libs/arm64-v8a -L /home/danielrodriguez/android-ndk-r18/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x -L /home/danielrodriguez/android-ndk-r18/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/aarch64-linux-android/lib -use-ld=gold  -module-cache-path '/home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/swift-test-results/aarch64-unknown-linux-android/clang-module-cache' -swift-version 4  -Xfrontend -ignore-module-source-info  -force-single-frontend-invocation -module-name PlaygroundSupport -emit-module-path /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp/PlaygroundSupport.swiftmodule -parse-as-library -c -o /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp/PlaygroundSupport.o /home/danielrodriguez/swift-source/swift/test/PlaygroundTransform/Inputs/SilentPCMacroRuntime.swift /home/danielrodriguez/swift-source/swift/test/PlaygroundTransform/Inputs/PlaygroundsRuntime.swift
: 'RUN: at line 4';   /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/bin/swiftc -target aarch64-unknown-linux-android -Xcc --sysroot=/home/danielrodriguez/android-ndk-r18/platforms/android-21/arch-arm64 -Xclang-linker --sysroot=/home/danielrodriguez/android-ndk-r18/platforms/android-21/arch-arm64 -tools-directory /home/danielrodriguez/android-ndk-r18/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/aarch64-linux-android/bin -I /home/danielrodriguez/android-ndk-r18/sysroot/usr/include -I /home/danielrodriguez/android-ndk-r18/sysroot/usr/include/aarch64-linux-android -L /home/danielrodriguez/android-ndk-r18/sources/cxx-stl/llvm-libc++/libs/arm64-v8a -L /home/danielrodriguez/android-ndk-r18/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x -L /home/danielrodriguez/android-ndk-r18/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/aarch64-linux-android/lib -use-ld=gold  -module-cache-path '/home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/swift-test-results/aarch64-unknown-linux-android/clang-module-cache' -swift-version 4  -Xfrontend -ignore-module-source-info  -Xfrontend -playground -o /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp/main -I=/home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp/PlaygroundSupport.o /home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/test-android-aarch64/PlaygroundTransform/Output/iuo.swift.tmp/main.swift
--
Exit Code: 1

Command Output (stderr):
--
/home/danielrodriguez/swift-source/build/my_android_aarch64/swift-linux-x86_64/lib/swift/android/aarch64/swiftrt.o:SwiftRT-ELF.cpp:function swift_image_constructor(): error: undefined reference to 'swift_addNewDSOImage'
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
<unknown>:0: error: link command failed with exit code 1 (use -v to see invocation)

--

********************

Testing Time: 0.37s
********************
Failing Tests (1):
    Swift(android-aarch64) :: PlaygroundTransform/iuo.swift

  Unexpected Failures: 1

2 warning(s) in tests
ERROR: command terminated with a non-zero exit status 1, aborting
```

Are we forgetting to link something?

buttaface commented:
I think that's another bug in the way cross-compilation from linux worked. That symbol `swift_addNewDSOImage` is defined in [`stdlib/public/runtime/ImageInspectionELF.cpp`](https://github.com/apple/swift/blob/37c6414e77185988567adad83ec00fc699f834c8/stdlib/public/runtime/ImageInspectionELF.cpp#L97), which is [only separated out into a different shared library if the static stdlib is also being built and if the host SDK is `LINUX`](https://github.com/apple/swift/blob/63ce2434377d8abbde881a80bbb6ffc0627055ad/stdlib/public/runtime/CMakeLists.txt#L92).

However, before this pull, [the CMake config for swiftCore would also wrongly link that added shared library anytime the static stdlib was being built and the primary variant SDK, ie the bogus alias for the host SDK mentioned above, was `LINUX`](https://github.com/apple/swift/pull/30170/files#diff-b697edde4923845275ebdcfaaa52222dL271), ie this was done for Android built on linux too.

So this bug covered up that bug till now, that one will have to be looked at and fixed, ie do we really only want that separate shared library only on linux or not?

buttaface commented:
I've added changes to the main commit for `swiftRuntime`, that get the static stdlib working when cross-compiling too. Using the following build-script command and backporting this patch to Swift 5.2, I was able to build the stdlib for both linux and Android with the official Swift 5.2 compiler for linux and verify that the linux stdlib tests still pass (I don't connect an Android device to the linux server I'm using, so I didn't run the Android stdlib tests with adb):
```
./swift/utils/build-script -R --no-assertions --build-subdir=52-release --skip-build-llvm --build-runtime-with-host-compiler --extra-cmake-options="-DSWIFT_INCLUDE_TOOLS=FALSE" --native-swift-tools-path=/home/butta/52-release/usr/bin/ --native-llvm-tools-path=/home/butta/52-release/usr/bin/ --native-clang-tools-path=/home/butta/52-release/usr/bin/ --android --android-ndk /data/ndk-r21/ --android-arch aarch64 --android-api-level 24 --android-icu-uc /data/data/com.termux/files/usr/lib/libicuuc.so --android-icu-uc-include /data/data/com.termux/files/usr/include/ --android-icu-i18n /data/data/com.termux/files/usr/lib/libicui18n.so --android-icu-i18n-include /data/data/com.termux/files/usr/include/ --android-icu-data /data/data/com.termux/files/usr/lib/libicudata.so --swift-install-components="stdlib;sdk-overlay" -j2 --install-swift --build-swift-static-stdlib --build-swift-static-sdk-overlay -T --skip-test-cmark --lit-args='--filter=stdlib/'
```
While this pull is aimed at cross-compiling the Android stdlib, it also makes the stdlib more cross-compilable to other platforms.

buttaface commented on stdlib/public/core/CMakeLists.txt:
```diff
@@ -230,11 +230,9 @@ set(SWIFTLIB_GYB_SOURCES
 set(swift_stdlib_compile_flags "${SWIFT_RUNTIME_SWIFT_COMPILE_FLAGS}")
 if(SWIFT_PRIMARY_VARIANT_SDK IN_LIST SWIFT_APPLE_PLATFORMS)
   list(APPEND swift_core_link_flags "-all_load")
-  list(APPEND swift_core_private_link_libraries icucore)
```
@vgorloff, this would try to link `icucore` for all cross-compiled stdlibs from a mac before this pull: would this cause problems for you when cross-compiling from mac to Android or [was it simply removed elsewhere](https://github.com/apple/swift/blob/d8de5bafb092a3a5d692c6dbd701cacff0a95c36/stdlib/cmake/modules/AddSwiftStdlib.cmake#L619)?

vgorloff replied:
Here is how configured SwiftBuilder on my side.

```sh
-D SWIFT_ANDROID_${arch.swiftArch}_ICU_UC=${icu.paths.installs}/lib/libicuucswift.so
-D SWIFT_ANDROID_${arch.swiftArch}_ICU_UC_INCLUDE=${icu.paths.sources}/source/common
-D SWIFT_ANDROID_${arch.swiftArch}_ICU_I18N=${icu.paths.installs}/lib/libicui18nswift.so
-D SWIFT_ANDROID_${arch.swiftArch}_ICU_I18N_INCLUDE=${icu.paths.sources}/source/i18n
-D SWIFT_ANDROID_${arch.swiftArch}_ICU_DATA=${icu.paths.installs}/lib/libicudataswift.so
```

Also I can't find library `icucore` in build output on my side.
<img width="340" alt="icu" src="https://user-images.githubusercontent.com/3200080/144494837-60cbfe4b-5642-4f9b-bf16-ded2578ec067.png">

buttaface replied:
Well, it should be moot once the ICU dependency is removed, #40340. 🎉

buttaface commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -149,7 +148,7 @@ if(SWIFT_BUILD_STATIC_STDLIB AND "${sdk}" STREQUAL "LINUX")
                              COMPONENT stdlib)
   add_custom_target(static_binary_magic ALL DEPENDS ${static_binary_lnk_file_list})
-  foreach(arch IN LISTS SWIFT_SDK_LINUX_ARCHITECTURES)
+  foreach(arch IN LISTS SWIFT_SDK_${sdk}_ARCHITECTURES)
```
I made this function as generic as possible, so that enabling it for other ELF platforms later would be easy.

compnerd replied:
This part of the change seems reasonable, and is probably a good thing to do.

buttaface commented:
@spevans, you added the static executable support in #5394 that I just modified, let me know what you think.

buttaface commented:
Added one last change to replace the last uses of `SWIFT_CONFIGURED_SDKS` with `SWIFT_SDKS`, so that the host SDK doesn't have to be built. Running the `build-script` command I pasted a couple days ago but replacing the last three testing-related flags with `--stdlib-deployment-targets=android-aarch64 --swift-primary-variant-sdk=ANDROID --swift-primary-variant-arch=aarch64`, this last change allows me to cross-compile the Swift stdlib for Android alone, without needing to build the host linux stdlib.

`SWIFT_SDKS` is used everywhere else in this repo and it is what should be used for generating these target SDK files. Otherwise, the above three flags break the CMake config, as build-script asks for only the Android stdlib to be built, but CMake will also try to configure the host SDK, linux in my case.

This pull is ready to go, just needs a review.

buttaface commented:
Updated this pull to move the generation of the static-stdlib linker file from `lib/Driver/CMakeLists.txt` to `stdlib/CMakeLists.txt`, otherwise it wouldn't be generated when compiling the static stdlib alone.

buttaface commented:
Ping, [this pull was used in its entirety](https://github.com/termux/termux-packages/blob/master/packages/swift/swift-cmake.patch.beforehostbuild#L89) to [cross-compile the stdlib of the pre-built toolchain for Android](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L87). It makes it possible to cross-compile the Swift stdlib without compiling the host compiler or stdlib, would be good to get this in before the 5.3 branch.

drexin commented:
@swift-ci smoke test

buttaface commented:
linux/mac worked, [Windows CI has been broken since last night because of a single unrelated test](https://ci-external.swift.org/job/swift-PR-windows/buildTimeTrend).

buttaface commented:
@gottesmm, requesting review again.

drodriguez commented:
The problems I experienced before in Android seems to be gone. Thanks.

I am not sure that I have the confidence to merge this. It touches many pieces of the build system. Seems to work, which is great.

@swift-ci please test Windows platform

buttaface commented:
Now that 5.3 has branched, should be okay to get this into master now, if there are worries it might break something and time is needed. All the first commit does is remove libatomic on other platforms, just like it has been on linux. All the second commit does is rewrite the way some linker flags and libraries are added, so that they won't be added to say the Android stdlib incorrectly just because it was cross-compiled from a linux host.

It changes nothing for host builds, which is mostly all that has been tried or tested with this codebase, applying all the same flags as before. The only change the second commit makes is switching from `SWIFT_CONFIGURED_SDKS` in a few places to `SWIFT_SDKS`, which is the right list of SDKs used everywhere else in this repo.

buttaface commented:
Rebased and this pull is now smaller, since Saleem has been silently merging pieces of it in his own pulls, over the two months this pull sat unreviewed.

@CodaFi, I mentioned this pull on the forum last month and [got some acknowledgement from @gottesmm](https://forums.swift.org/t/separate-cmake-process-into-compiler-and-stdlib/35392/6), though he says he has bigger plans in this vein this summer. I see no reason why that should preclude merging this small working pull in the meantime, which is already being used when compiling [the Swift compiler for Android](https://bintray.com/termux/termux-packages-24/swift).

compnerd commented on stdlib/public/core/CMakeLists.txt:
```diff
@@ -247,35 +245,40 @@ else()
-      ${SWIFT_${SWIFT_PRIMARY_VARIANT_SDK}_${SWIFT_PRIMARY_VARIANT_ARCH}_ICU_I18N})
+endif()
+
+function(swift_core_private_libraries sdk arch libraries_var_name)
```
I think that this is the going in the exact opposite direction of where we want to go.  We need to start using `find_library` and imported targets to create the dependencies rather than hard coded names and magic flags.

buttaface replied:
It is not clear why you placed this review comment here: this reorganizes a chunk of the following CMake config as a CMake function and is orthogonal to the issue of "imported targets" vs "magic flags." Maybe you meant to place it below, where the flags are?

As for the way those libraries and flags are added below, that predates this pull and has nothing to do with me. I realize you'd like to see it modernized, but that's not what this pull is offering: I simply refactor the way CMake calls the pre-existing "magic flags," so they can be used when cross-compiling too.

compnerd commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -213,7 +213,7 @@ add_swift_target_library(swiftImageRegistrationObjectCOFF
                   INSTALL_IN_COMPONENT none)
 
-foreach(sdk ${SWIFT_CONFIGURED_SDKS})
+foreach(sdk ${SWIFT_SDKS})
```
This change seems principally incorrect: only the targets that have been setup to be built are the ones that we should be setting up targets for.

buttaface replied:
No, [as explained when I added this change last month](https://github.com/apple/swift/pull/30170#issuecomment-607065583), `SWIFT_SDKS` is the right one because it is possible for the developer running `build-script` to ask for only a subset of the configured SDKs to be built with `--stdlib-deployment-targets`, which is where `SWIFT_SDKS` comes from. It is why `SWIFT_SDKS` is used everywhere else in this repo, with [`SWIFT_CONFIGURED_SDKS` now only relegated to making sure the passed-in `SWIFT_SDKS` is a subset of it](https://github.com/apple/swift/blob/2b2b5b8f081fcd2f2f0a3609276f920f96c1f7b4/CMakeLists.txt#L797).

buttaface commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -379,11 +379,6 @@ function(_add_target_variant_link_flags)
-    # link against the ICU libraries
-    list(APPEND link_libraries
-      ${SWIFT_ANDROID_${LFLAGS_ARCH}_ICU_I18N}
-      ${SWIFT_ANDROID_${LFLAGS_ARCH}_ICU_UC})
```
Missed this tweak from the rebase yesterday, because it was moved to a different file upstream, added it back now.

buttaface commented:
Getting back to this after more than a year, rebased and tweaked it for how things are done now.

Let me explain more broadly why this pull is needed:

Saleem moved a bunch of target flags from the central `add_swift_target_library()` to this `swiftCore` library CMake configuration in #21707, presumably because those flags were only needed by this library. That works fine if you're only building the host stdlib, but breaks if you're building multiple stdlibs/SDKs alongside the host stdlib, because previously `add_swift_target_library()` added these flags on a per-SDK basis, but `stdlib/public/core/CMakeLists.txt` can only apply flags much more broadly.

To remedy that, I kept the logic adding these link flags in `stdlib/public/core/CMakeLists.txt` but made it a function `swift_core_private_libraries()` that is still invoked in `add_swift_target_library()`, so these flags are still added correctly on a per-SDK basis.

This fixes [the command we give people to build the Android SDK](https://github.com/apple/swift/blob/main/docs/Android.md#2-building-the-swift-stdlib-for-android), which has been [reported as broken on the forum](https://forums.swift.org/t/building-the-swift-stdlib-for-android-linking-icu-libraries-fails/44720).

@compnerd, let me know what you think of the latest iteration of this pull.

@edymtt, you know these CMake issues well, would appreciate your input.

buttaface commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -148,6 +123,30 @@ add_swift_target_library(swiftImageRegistrationObjectCOFF
                   INSTALL_IN_COMPONENT none)
 
 foreach(sdk ${SWIFT_SDKS})
+  if(SWIFT_BUILD_STATIC_STDLIB AND "${SWIFT_SDK_${sdk}_OBJECT_FORMAT}" STREQUAL "ELF")
```
Rather than only generate this link file if the host SDK is linux, as was done above, create it for all ELF SDKs in the list of wanted `SWIFT_SDKS`.
